Avoid unnecessary restore_file calls and option to resume at specific path

tvogel opened the pull request:
I made two patches:

- the first checks whether the file is already at the same revision (or both revisions are "deleted") and avoids the client.restore() call in that case in order to speed up the script
- the second allows to optionally specify a path to resume at: I often encounter the dropbox server to bail out with [500] u'Internal Server Error' and it's really helpful to resume there.

Hope you find it useful!

clark800 commented:
Thanks for the pull request.

For the first patch, isn't this functionality already there? https://github.com/clark800/dropbox-restore/blob/master/restore.py#L50-L54

For the second patch, I don't like the idea of trusting the API to always return results in the same order since I don't think the documentation guarantees this. You could sort locally, but I would prefer if Dropbox just fixes there 500 errors rather than complicating the script. Did you submit an issue with them?

tvogel commented:
Hi!

Am 19.01.2015 um 00:14 schrieb Chris Clark:
> For the first patch, isn't this functionality already there?
> https://github.com/clark800/dropbox-restore/blob/master/restore.py#L50-L54

No, because it will call restore in the following unnecessary cases:

- the file has been restored to the same revision before but after the
  cut-off time
- the file was deleted at the cut-off time but restored and deleted
  again after the cut-off time

> For the second patch, I don't like the idea of trusting the API to
> always return results in the same order since I don't think the
> documentation guarantees this. You could sort locally, but I would
> prefer if Dropbox just fixes there 500 errors rather than complicating
> the script. Did you submit an issue with them?

I never saw a folder listing returned in anything else than alphabetical
order. Actually, I just adjusted the script to retry when receiving a
500 in client.restore() which could also workaround the resume issue.
Sorting is a good idea. I just sent a report to the developer support.

Anyway, being able to resume is useful to have, I think.

Regards,

Tilman

clark800 commented:
Nice catch on that first point, thanks! I wrote it a little differently because again I didn't want to assume anything that the API docs don't guarantee, in this case that the revisions are sorted. Also, I am going to leave out the deletion check because technically in the case you describe, you should still restore to get the exact metadata since the rev field is different (even though from the user perspective it doesn't make a difference).

Here is the new pull request, let me know if this works for you: https://github.com/clark800/dropbox-restore/pull/24/files

If you make a separate pull request with the retry on 500 error code, I'll take a look at that.

tvogel commented:
Am 19.01.2015 um 04:34 schrieb Chris Clark:
> Nice catch on that first point, thanks! I wrote it a little differently
> because again I didn't want to assume anything that the API docs don't
> guarantee, in this case that the revisions are sorted. Also, I am going
> to leave out the deletion check because technically in the case you
> describe, you should still restore to get the exact metadata since the
> rev field is different (even though from the user perspective it doesn't
> make a difference).

I think that restore will in fact not reuse the previous rev but instead
generate a new one which has the same contents as the previous rev. So,
the rev itself cannot be restored.

To me, it's important to get the content state back in shape with least
effort (least API calls). Because we cannot really roll back history
(and probably it's also not intended), that's enough for me. What other
metadata are you concerned of? Even the last modifier will be reset to
the user executing the script. There is no way to impersonate the last
modifier before the cut-off.

Regards,

Tilman

clark800 commented:
> I think that restore will in fact not reuse the previous rev but instead
> generate a new one which has the same contents as the previous rev. So,
> the rev itself cannot be restored.

If that were the case, then the first optimization you suggested would have no effect because the rev field would always be different.

I incorporated the retry functionality, but I did it a little differently to make it more readable and catch errors on any API call in the restore_file function. Here is the latest: https://github.com/clark800/dropbox-restore/pull/24
